### ISSUE TITLE
Multi-Word Correction and Partial Substitutions

### ISSUE THREAD

**Issue by Softwaerewolf** [futo-org/android-keyboard#218]
Currently it seems suggestions strongly prefer offering matching single words or hyphenated words.

My most common typing mistake is missing space and hitting "v" instead. In Swiftkey it's suggestions substitute these incorrect characters for a space or at least offer multiple word corrections.

Although possible, FUTO keyboard seems extremely adverse to multi-word suggestions. It will do it if you select a bad correction and then undo and replace it with your preferred correction.

For example:
Helloeorld =(select)=> HelloWorld =(manually delete and correct)=> Hello world

Now it will offer "Hello world" as a suggestion (not guaranteed in all circumstances). But it should do this automatically from the start.

The behavior is actually even more erratic.

For example:
"Thelazyfox" offers "Tulztocracy", but it does seem aware that these could be 3 separate words if I simply don't select a correction, then move my cursor to the beginning of the concatinated word. Now it offers "The lazy fox", but if I type out the same concatinated words again it does not.

Additionally, "thevlazybfox" should detect and offer the same three word correction. It does not under any circumstances.

Finally, but relatedly, partial substitutions. Where some part of a word is known but the rest isn't or is ambiguous.

For example
"Sixallles" could be "Six allies", " Six apples" or "Six alleles" but there is a high confidence that the first word is "Six".

Far future:
There are knock-on benefits to partial corrections which will be necessary for certain languages such as Pinyin input for Chinese where words share pinyin and sometimes characters.

For (probably bad) example:
舊時 jiushi - old times
舊識 jiushi - old friend
舅氏 jiushi - uncle

Knowing which is the first character informs which is the most likely second character.

Typing also commonly involves typing entire sentences with just initials, eg. "Ndmzshenm" becomes "你的名字什么" or "Ni De MingZi Shen Me"

Note, I am not Mandarin speaker, my information is all observational.

**Comment by Softwaerewolf** [comment]
Although this suggestion feels overloaded, I have put additional information about why this functionality should be pursued as a matter of course because of impacts it has in relation to internationalisation for certain languages.

The same functionality that is a "nice to have" in English and Latin and similar languages becomes "essential" in relation to Chinese and possibly Japanese and other languages.

**Comment by Philippe-Choquette** [comment]
Yes!!! I'm a bit of a lazy typer and tend to accidentally hit a character instead of the spacebar (letter "v" usually), and because FUTO keyboard's auto-correct algorithm has literally no ability to understand this, I always have to backspace to correct it.

**Comment by Softwaerewolf** [comment]
Reverted the erroneous renaming

**Comment by abb128** [comment]
This was improved in v0.1.24.2

**Comment by Softwaerewolf** [comment]
Retesting this in `0.1.27` and it is no better.

There does not appear to be any word segmentation done.

There exists open source implementations available (eg. https://github.com/dchentech/wordsegment) that can be used or used as reference.

As dictionary files are already included as part of the autocorrect feature, a corpus of words for segmentation is already available.

Once segmentation is performed, each segment needs only go through spelling correction independently to resolve a majority of these anomalies; "sixallles" => "six allles" => "six apples".

Word segmentation can be comfortably limited to three words.

**Comment by abb128** [comment]
@Softwaerewolf It is better and there are multi word suggestions being done up to two words. For example your original example of "helloeorld" will give you the option of "hello world". Currently the algorithm only considers space substitutions for a lower row key (e.g. accidentally hitting n instead of space) and not deletions, so "sixnapples" will correct to "six apples", and "sixallles" for me gives "so allows". The hitbox is also bigger for the spacebar to reduce substitution happening after you write "six". If you have any specific improvement requests you can open a new issue to keep track of it better.
